### ISSUE TITLE
DRLearner's score is wrong when final model returns an array instead of a vector

### ISSUE THREAD

**Issue by kbattocchi** [microsoft/EconML#420]
The logic for computing the score here:

https://github.com/microsoft/EconML/blob/05fb59595244dc07d3557e75fb2c7a52407921bb/econml/dr/_drlearner.py#L164-L167

is wrong if the final model's `predict` function returns a one-column 2D array even when `fit` was called with a 1D vector target (or vice versa), because in that case we end up broadcasting a vector from indexing into `Y_pred` against an array from `predict(X)` when subtracting.

Note that we can't just fix this by reshaping one of the arrays in the subtraction operation to the shape of the other, because if X was None then we really do want to broadcast because the result of `predict(X)` will be of size 1.

Here's a simple repro: the class L is just a LinearRegression that reshapes the output of `predict` - when `shape1` and `shape2` match, then the resulting score is correct, but when it doesn't the score is wrong (and generally higher).

```
controls = np.random.uniform(-1, 1, size=(10000, 2))
T = np.random.binomial(2, scipy.special.expit(controls[:, 0]))
sigma = 0.01
y = (1 + .5 * controls[:, 0]) * T + controls[:,0] + np.random.normal(0, sigma, size=(10000,))

cv = [(np.arange(5000), np.arange(5000,10000)),
      (np.arange(5000,10000), np.arange(5000))]

class L:
    def __init__(self, shape):
        self.model = LinearRegression()
        self.shape = shape

    def fit(self, X, y):
        self.model.fit(X, y)
        return self

    def predict(self, X):
        return self.model.predict(X).reshape(*self.shape)

scores = {}

for shape1 in (-1,),(-1,1):
  for shape2 in (-1,),(-1,1):
    y = y.reshape(shape2)
    est = DRLearner(model_propensity=LogisticRegression(),
                  model_regression=Lasso(),
                  model_final=L(shape1),
                  cv=cv).fit(y, T, X=controls)
    scores[(len(shape1), len(shape2))] = est.score(y, T, X=controls)
scores
```

**Comment by heimengqi** [comment]
You are right there is a broadcasting issue, but It's in another place than the code you link to above
https://github.com/microsoft/EconML/blob/05fb59595244dc07d3557e75fb2c7a52407921bb/econml/dr/_drlearner.py#L160-L162

It actually happens when `multitask_model_final=True` and Y is a 2D array (either (m,1) or (m,d_y)). In the final wrapper during fit, we squeeze Y from (m,1,d_t) to (m, d_t) in order to input as 2D array into `model_cate`, then during predict, we add the y axis back. However, under the score function we forgot to do that, that causes shape mismatch.

It will be resolved by #419

**Comment by heimengqi** [comment]
After looking at the build error, yes the code you link to has another issue, it's coming form the RegressionForest, I will fix it.
